src/editor/actions/server-save-actions.js

```diff
@@ -0,0 +1,155 @@
+    ) {
+      status = "ERROR_OUT_OF_DATE";
+    } else {
+      status = "ERROR_GENERAL";
+    }
+  } else {
+    status = "OK";
+  }
+  return {
+    type: "SET_SERVER_SAVE_STATUS",
+    status
+  };
+}
+
+export function createNewNotebookOnServer() {
+  return async (dispatch, getState) => {
+    const state = getState();
+    const originalNotebookId = getNotebookID(state);
+    try {
+      const notebook = await createNotebookRequest(
+        state.title,
+        state.jsmd,
+        originalNotebookId ? { forked_from: originalNotebookId } : {}
+      );
+      window.history.replaceState({}, "", `/notebooks/${notebook.id}/`);
+      dispatch({ type: "ADD_NOTEBOOK_ID", id: notebook.id });
+      dispatch({
+        type: "NOTEBOOK_SAVED",
+        newRevisionId: notebook.latest_revision.id
+      });
+      dispatch(setServerSaveStatus(undefined));
+
+      // remove previous autosave if we've successfully saved.
+      clearLocalAutosave(state);
+
+      // update owner info, so we know that we can save
+      if (originalNotebookId) {
+        dispatch({
+          type: "SET_NOTEBOOK_OWNER_IN_SESSION",
+          owner: state.userData
+        });
+      }
+    } catch (e) {
+      dispatch(setServerSaveStatus(e));
+
+      // re-throw so other consumers can handle the error (e.g. to pop up
+      // an app message)
+      throw e;
+    }
+  };
+}
+
+export function saveNotebookToServer(forceSave = false) {
+  return async (dispatch, getState) => {
+    const state = getState();
+    const notebookId = getNotebookID(state);
+    const notebookInServer = Boolean(notebookId);
+    try {
+      // no notebook exists yet on server, need to create one
+      if (!notebookInServer) {
+        dispatch(createNewNotebookOnServer());
+        return;
+      }
+      // save existing revision
+      const newRevision = await updateNotebookRequest(
+        notebookId,
+        forceSave ? undefined : getRevisionID(state),
+        state.title,
+        state.jsmd
+      );
+      dispatch({ type: "NOTEBOOK_SAVED", newRevisionId: newRevision.id });
+      dispatch(setServerSaveStatus(undefined));
+
+      // remove previous autosave if we've successfully saved.
+      clearLocalAutosave(state);
+    } catch (e) {
+      dispatch(setServerSaveStatus(e));
+      // re-throw so other consumers can handle the error (e.g. to pop up
+      // an app message)
+      throw e;
+    }
+  };
+}
+
+export function checkNotebookIsBasedOnLatestServerRevision() {
+  return async (dispatch, getState) => {
+    const state = getState();
+    const notebookId = getNotebookID(state);
+    if (!notebookId) {
+      // no notebook id assigned yet, so not possible to be
+      // out of date
+      return;
+    }
+    try {
+      const notebookData = await getNotebookRequest(notebookId);
+      dispatch({
+        type: "UPDATE_NOTEBOOK_REVISION_IS_LATEST",
+        revisionIsLatest:
+          notebookData.latest_revision.id === getRevisionID(state)
```

Comment:
In testing I just realized that this function will confusingly say the notebook is out of date even if the content is the exact same as that on the server (but the revision is different). This seems unlikely to happen, but seems worth fixing since it's quite easy (just need to expand on this line a bit).